### ISSUE TITLE
enhancement: agent should not abort if it cannot connect to BIG-IPs on boot

### ISSUE THREAD

**Issue by alonsocamaro** [F5Networks/f5-openstack-agent#395]
#### OpenStack Release

Mitaka
#### Description

Hi

Currently the agent startup script aborts after retrying to connect to the BIG-IPs for few seconds (se sample log below). This is not suitable for a production environment.

This could be caused because:
-   Openstack startup is asynchronous and takes long time or 
-   The BIG-IPs are not available on agent startup.

A better behavior would be that the agent tries to reconnect to the BIG-IPs periodically. This should cover the two cases above.

```

2016-10-16 19:48:03.074 16406 INFO neutron.common.config [-] /usr/bin/f5-oslbaasv2-agent version 8.1.1
2016-10-16 19:48:03.739 16406 INFO f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver [req-e73b6639-b832-4344-9b5c-0402f9d3fd12 - - - - -] Opening iControl connection to admin @ 10.128.5.237
2016-10-16 19:48:06.745 16406 ERROR f5_openstack_agent.lbaasv2.drivers.bigip.icontrol_driver [req-e73b6639-b832-4344-9b5c-0402f9d3fd12 - - - - -] Could not communicate with all iControl devices: HTTPSConnectionPool(host='10.128.5.237', port=443): Max retries exceeded with url: /mgmt/tm/sys/ (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f7757a086d0>: Failed to establish a new connection: [Errno 113] EHOSTUNREACH',))
2016-10-16 19:48:11.748 16406 CRITICAL neutron [req-e73b6639-b832-4344-9b5c-0402f9d3fd12 - - - - -] ConnectionError: HTTPSConnectionPool(host='10.128.5.237', port=443): Max retries exceeded with url: /mgmt/tm/sys/ (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f7757a086d0>: Failed to establish a new connection: [Errno 113] EHOSTUNREACH',))
2016-10-16 19:48:11.748 16406 ERROR neutron Traceback (most recent call last):
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/bin/f5-oslbaasv2-agent", line 9, in <module>
2016-10-16 19:48:11.748 16406 ERROR neutron     load_entry_point('f5-openstack-agent==9.1.0', 'console_scripts', 'f5-oslbaasv2-agent')()
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5_openstack_agent/lbaasv2/drivers/bigip/agent.py", line 69, in main
2016-10-16 19:48:11.748 16406 ERROR neutron     mgr = manager.LbaasAgentManager(cfg.CONF)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5_openstack_agent/lbaasv2/drivers/bigip/agent_manager.py", line 238, in __init__
2016-10-16 19:48:11.748 16406 ERROR neutron     self._load_driver(conf)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5_openstack_agent/lbaasv2/drivers/bigip/agent_manager.py", line 286, in _load_driver
2016-10-16 19:48:11.748 16406 ERROR neutron     self.conf)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/oslo_utils/importutils.py", line 44, in import_object
2016-10-16 19:48:11.748 16406 ERROR neutron     return import_class(import_str)(*args, **kwargs)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5_openstack_agent/lbaasv2/drivers/bigip/icontrol_driver.py", line 371, in __init__
2016-10-16 19:48:11.748 16406 ERROR neutron     self.connect_bigips()
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5_openstack_agent/lbaasv2/drivers/bigip/icontrol_driver.py", line 387, in connect_bigips
2016-10-16 19:48:11.748 16406 ERROR neutron     self._init_bigips()
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5_openstack_agent/lbaasv2/drivers/bigip/icontrol_driver.py", line 509, in _init_bigips
2016-10-16 19:48:11.748 16406 ERROR neutron     first_bigip = self._open_bigip(self.hostnames[0])
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5_openstack_agent/lbaasv2/drivers/bigip/icontrol_driver.py", line 542, in _open_bigip
2016-10-16 19:48:11.748 16406 ERROR neutron     self.conf.icontrol_password)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5/bigip/__init__.py", line 65, in __init__
2016-10-16 19:48:11.748 16406 ERROR neutron     self._get_tmos_version()
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/f5/bigip/__init__.py", line 82, in _get_tmos_version
2016-10-16 19:48:11.748 16406 ERROR neutron     response = connect.get(base_uri)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/icontrol/session.py", line 216, in wrapper
2016-10-16 19:48:11.748 16406 ERROR neutron     response = method(self, REST_uri, **kwargs)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/icontrol/session.py", line 329, in get
2016-10-16 19:48:11.748 16406 ERROR neutron     return self.session.get(uri, **kwargs)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 480, in get
2016-10-16 19:48:11.748 16406 ERROR neutron     return self.request('GET', url, **kwargs)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 468, in request
2016-10-16 19:48:11.748 16406 ERROR neutron     resp = self.send(prep, **send_kwargs)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 576, in send
2016-10-16 19:48:11.748 16406 ERROR neutron     r = adapter.send(request, **kwargs)
2016-10-16 19:48:11.748 16406 ERROR neutron   File "/usr/lib/python2.7/dist-packages/requests/adapters.py", line 437, in send
2016-10-16 19:48:11.748 16406 ERROR neutron     raise ConnectionError(e, request=request)
2016-10-16 19:48:11.748 16406 ERROR neutron ConnectionError: HTTPSConnectionPool(host='10.128.5.237', port=443): Max retries exceeded with url: /mgmt/tm/sys/ (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f7757a086d0>: Failed to establish a new connection: [Errno 113] EHOSTUNREACH',))
2016-10-16 19:48:11.748 16406 ERROR neutron 

```
#### Agent Version

9.1.0
#### Operating System

Mirantis 9

**Comment by alonsocamaro** [comment]
Note that this is already handled when using the systemd scripts which are included in the agent packages, these restart the agent if it has exited, see the following examples:

https://github.com/F5Networks/f5-openstack-agent/blob/liberty/lib/systemd/system/f5-openstack-agent.service
 
https://github.com/F5Networks/f5-openstack-agent/blob/mitaka/lib/systemd/system/f5-openstack-agent.service

**Comment by richbrowne** [comment]
Fixed by #730 and #1085
